devito/at_controller.py

```diff
@@ -91,71 +92,69 @@ def auto_tune_blocks(self, minimum=5, maximum=20):
+        block_list = sorted(self._filter(block_list))
 
-        logger.info("Total number of block size permutations = %d" % len(block_list))
-        if maximum - minimum > 2:  # No point to estimate if diff is below 2.
-            self._estimate_run_time(minimum, maximum, len(block_list))
+        # populate output arrays with random values, to make sure that loads
+        # and stores are performed
+        for param in self.op.output_params:
+            param._data = np.random.rand(*param.data.shape).astype(self.op.dtype)
+
+        info_at("Number of block sizes that will be attempted: %d" % len(block_list))
 
         # runs function for each block_size
+        times = []
+        self.op.propagator.cache_blocking = list(block_list[0])
         for block in block_list:
-            self.op.propagator.block_sizes = block
+            self.op.propagator.block_sizes = list(block)
             times.append((block, self.get_execution_time()))
 
         # sorts the list of tuples based on time
         times = sorted(times, key=lambda element: element[1])
 
-        logger.info("Auto tuning using brute force complete")
-        logger.info("Estimated runtime for %s and %d time steps: %f hours" %
-                    (self.op.getName(), self.nt_full,
-                     self.nt_full * times[0][1] / (at_nt * 3600)))
+        info_at("Finish.")
+        info_at("Estimated runtime for %s and %d time steps: %f hours" %
+                (self.op.getName(), self.nt_full,
+                 self.nt_full * times[0][1] / (at_nt * 3600)))
 
         self._write_block_report(times)  # writes the report
 
-    def _estimate_run_time(self, minimum, maximum, n):
-        """Estimates run time for auto tuning
-
-        :param minimum: int - minimum tune range
-        :param maximum: int - maximum tune range
-        :param n: number of permutations that are run
+    def _filter(self, block_list):
         """
-        timing_run = 0  # estimating run time
-        for i in range(0, 5):
-            logger.info('Estimating auto-tuning runtime...sample %d/5' % (i + 1))
+        Filter off some block sizes to speed up autotuning. The current
+        heuristic is: ::
 
-            blocks = []
-            blocks += [random.randrange(minimum, maximum) if block else None
-                       for block in self.op.propagator.block_sizes]
+            * block sizes that are not a multiple of 2 are ditched;
+            * only square blocks are retained.
 
-            self.op.propagator.block_sizes = blocks
-            timing_run += self.get_execution_time()
-
-        logger.info("Estimated runtime: %f minutes." % float(timing_run / 5 * n / 60))
+        If no block sizes match the heuristic, then the original list is returned.
+        """
+        filtered_list = [b for b in block_list if all(i % 2 == 0 for i in b if i)]
+        filtered_list = [b for b in filtered_list if all(i == b[0] for i in b if i)]
+        return filtered_list or block_list
```

Comment:
Looks great for now, but I wonder if we should just generate the block list according to some heuristic, instead of creating a list and then trimming it. This is not a show-stopper though, just a thought.